### PR TITLE
[TRAFODION-2308]JDBC T4 support read LOB

### PR DESCRIPTION
**Pull request by xwq**

### CODE REVIEW

**Comment by traflm** [comment]
@kevinxu021 , @arvind-narain  could one of you review the code?

**Comment by xwq** [comment]
I think ruoyu is doing the review these days.

File: core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp [hunk 1]
```diff
@@ -6378,6 +6378,124 @@ odbc_SQLSrvr_FetchPerf_sme_(
...
+
+    try
+    {
+        SQL_EXEC_SetParserFlagsForExSqlComp_Internal(INTERNAL_QUERY_FROM_EXEUTIL);
+
+        short retcode = QryLobExtractSrvrStmt->ExecDirect(NULL, LobExtractQuery, EXTERNAL_STMT, TYPE_CALL, SQL_ASYNC_ENABLE_OFF, 0);
+
+        if (retcode == SQL_ERROR)
+        {
+            ERROR_DESC_def *p_buffer = QryLobExtractSrvrStmt->sqlError.errorList._buffer; strncpy(RequestError, p_buffer->errorText, sizeof(RequestError) - 1);
```
Comment (ryzuo) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: Please start a new line for this strncpy.

File: core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp [hunk 2]
```diff
@@ -6378,6 +6378,124 @@ odbc_SQLSrvr_FetchPerf_sme_(
...
+                ODBCMX_SERVER,
+                srvrGlobal->srvrObjRef,
+                1,
+                "Exception in executing EXTRACT LOBLENGTH");
+
+        exception_->exception_nr = odbc_SQLsrvr_ExtractLob_ParamError_exn_;
+        exception_->u.ParamError.ParamDesc = SQLSVC_EXCEPTION_EXECDIRECT_FAILED;
+    }
+
+    lobDataValue = new IDL_char(lobDataLen + 1);
```
Comment (ryzuo) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: Please do check the availability of the mem after new operation.

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/ExtractLobMessage.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+package org.trafodion.jdbc.t4;
+
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.UnsupportedCharsetException;
+
+class ExtractLobMessage {
+
+	static final short LOB_EXTRACT_LEN                 = 0;
+	static final short LOB_EXTRACT_BUFFER              = LOB_EXTRACT_LEN + 1;
+	static final short LOB_EXTRATC_BOTH_LEN_AND_BUFFER = LOB_EXTRACT_BUFFER + 1;
```
Comment (DaveBirdsall) [inline core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/ExtractLobMessage.java]: Did you mean, LOB_EXTRACT_BOTH_LEN_AND_BUFFER?

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/ExtractLobReply.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+class ExtractLobReply {
+	int totalErrorLength;	
+	SQLWarningOrError[] errorList;
+
+	odbc_SQLSvc_ExtractLob_exc_ m_p1;
+	String m_p2;
+	SQLItemDescList_def m_p3;
+	ERROR_DESC_LIST_def m_p4;
+	String proxySyntax = "";
```
Comment (DaveBirdsall) [inline core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/ExtractLobReply.java]: I don't see members m_p2, m_p3, m_p4 and proxySyntax used anywhere. Do we need them?

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/ExtractLobReply.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+	int lobDataLen = 0;
+
+	byte[] lobDataValue = null;
```
Comment (DaveBirdsall) [inline core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/ExtractLobReply.java]: It doesn't look like lobDataLen or lobDataValue are used outside the constructor. Do they need to be members or could they be locally-declared variables?

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Blob.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		if (isFreed_)
+			throw TrafT4Messages.createSQLException(null, null, "lob_has_been_freed", null);
+	}
+
+	public long length() throws SQLException {
+		testAvailability();
+		return data_.length;
+	}
+
+	public void free() throws SQLException {
```
Comment (DaveBirdsall) [inline core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Blob.java]: Looks like this method doesn't throw any exceptions at all.

File: core/conn/odbc/src/odbc/nsksrvr/Interface/marshalingsrvr_srvr.cpp [hunk 7]
```diff
@@ -2300,3 +2300,89 @@ MxoSrvr_ValidateToken_param_res_(
...
+    if (lobDataValue != NULL)
+    {
+        wlength += lobDataLen;
+    }
+    wlength += lobDataLen;
+
+    // update the length of message
+    message_length = wlength;
+
+    buffer = pnode->w_allocate(message_length);
```
Comment (arvind-narain) [inline core/conn/odbc/src/odbc/nsksrvr/Interface/marshalingsrvr_srvr.cpp]: You may want to return CEE_ALLOCFAIL if buffer is NULL.

File: core/conn/odbc/src/odbc/nsksrvr/Interface/marshalingsrvr_srvr.cpp [hunk 8]
```diff
@@ -2300,3 +2300,89 @@ MxoSrvr_ValidateToken_param_res_(
...
+            break;
+
+        case odbc_SQLsrvr_ExtractLob_InvalidConnection_exn_:
+        case odbc_SQLSrvr_ExtractLob_SQLInvalidhandle_exn_:
+            break;
+
+        default:
+            break;
+    }
+    IDL_long_copy((IDL_long *)&lobDataLen, curptr);
```
Comment (arvind-narain) [inline core/conn/odbc/src/odbc/nsksrvr/Interface/marshalingsrvr_srvr.cpp]: At this point you could check for buffer overflow ( currptr > buffer + message_length )

File: core/conn/odbc/src/odbc/nsksrvr/Interface/odbcs_srvr_res.cpp [hunk 9]
```diff
@@ -642,6 +642,41 @@ odbc_SQLSrvr_EndTransaction_ts_res_(
...
+
+    CEE_status sts = CEE_SUCCESS;
+
+    IDL_short error;
+    CEERCV_IOMessage_reply_seq_ reply;
+
+    char * buffer = NULL;
+    UInt32 message_length = 0;
+
+    odbc_SQLsrvr_ExtractLob_param_res_(
```
Comment (arvind-narain) [inline core/conn/odbc/src/odbc/nsksrvr/Interface/odbcs_srvr_res.cpp]: If odbc_SQLsrvr_ExtractLob_param_res_ does return error ( CEE_ALLOCFAIL - previous comment ) then store the return status in sts here and check for the same.

File: core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp [hunk 10]
```diff
@@ -6378,6 +6378,124 @@ odbc_SQLSrvr_FetchPerf_sme_(
...
+  , /* In    */ IDL_string lobHandle
+  , /* Out   */ IDL_long_long &lobDataLen
+  , /* Out   */ IDL_char *& lobDataValue
+  )
+{
+    char LobExtractQuery[1000];
+    char RequestError[200] = {0};
+    SRVR_STMT_HDL  *QryLobExtractSrvrStmt = NULL;
+
+    if ((QryLobExtractSrvrStmt = getSrvrStmt("", TRUE)) == NULL)
```
Comment (arvind-narain) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: Could the corresponding statements for extract be prepared once with parameters and then executed ?

File: core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp [hunk 11]
```diff
@@ -6378,6 +6378,124 @@ odbc_SQLSrvr_FetchPerf_sme_(
...
+                srvrGlobal->nskProcessInfo.processId,
+                ODBCMX_SERVER,
+                srvrGlobal->srvrObjRef,
+                1,
+                "Exception in executing EXTRACT LOBLENGTH");
+
+        exception_->exception_nr = odbc_SQLsrvr_ExtractLob_ParamError_exn_;
+        exception_->u.ParamError.ParamDesc = SQLSVC_EXCEPTION_EXECDIRECT_FAILED;
+    }
+
```
Comment (arvind-narain) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: Would you like to check for exception_->exception_nr == 0 here ?

File: core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp [hunk 12]
```diff
@@ -6378,6 +6378,124 @@ odbc_SQLSrvr_FetchPerf_sme_(
...
+                    srvrGlobal->srvrObjRef,
+                    1,
+                    "Exception in executing EXTRACT LOBTOBUFFER");
+
+            exception_->exception_nr = odbc_SQLsrvr_ExtractLob_ParamError_exn_;
+            exception_->u.ParamError.ParamDesc = SQLSVC_EXCEPTION_EXECDIRECT_FAILED;
+        }
+    }
+}
+
```
Comment (arvind-narain) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: return at this point ?

File: core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp [hunk 13]
```diff
@@ -6378,6 +6378,124 @@ odbc_SQLSrvr_FetchPerf_sme_(
...
+    }
+
+    snprintf(LobExtractQuery, sizeof(LobExtractQuery), "EXTRACT LOBLENGTH(LOB'%s') LOCATION %Ld", lobHandle, (Int64)&lobDataLen);
+
+    try
+    {
+        SQL_EXEC_SetParserFlagsForExSqlComp_Internal(INTERNAL_QUERY_FROM_EXEUTIL);
+
+        short retcode = QryLobExtractSrvrStmt->ExecDirect(NULL, LobExtractQuery, EXTERNAL_STMT, TYPE_CALL, SQL_ASYNC_ENABLE_OFF, 0);
+
```
Comment (arvind-narain) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: Do please first get the parser flag, save it and after each ExecDirect set it back to the saved parser flag.

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Blob.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+	public byte[] getBytes(long pos, int length) throws SQLException {
+		testAvailability();
+
+		byte[] buf = null;
+		int startPos = (int)pos - 1;
+		int endPos = startPos + length;
+
+		if((startPos > data_.length) || (startPos < 0) || (length < 0))
+			throw TrafT4Messages.createSQLException(this.connection_.props_, this.connection_.getLocale(),
+					"out_of_lob_bound_msg", null);
```
Comment (kevinxu021) [inline core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Blob.java]: should be "out_of_log_bound"

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Blob.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+	public int setBytes(long pos, byte[] bytes, int offset, int len) throws SQLException {
+		testAvailability();
+		OutputStream out = setBinaryStream(pos);
+
+		try {
+			out.write(bytes, offset, len);
+		} catch (IOException e) {
+			throw TrafT4Messages.createSQLException(this.connection_.props_, this.connection_.getLocale(),
+					"blob_io_error", null);
```
Comment (kevinxu021) [inline core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Blob.java]: null to e.getMessage()

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Clob.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+						connection_.ic_);
+				LogicalByteArray rbuffer = t4connection.getReadBuffer(TRANSPORT.SRVR_API_EXTRACTLOB, wbuffer);
+				ExtractLobReply reply = new ExtractLobReply(rbuffer, connection_.ic_);
+
+				this.data_ = new String(reply.lobDataValue, "UTF-8");
+				isFreed_ = false;
+			}
+		}
+		catch (Exception e) {
+
```
Comment (kevinxu021) [inline core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Clob.java]: log this exception

File: core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp (hunk 13 above)
Comment (xwq) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: Hi, Arvind.

I'm not quite understand the parser flag. I'm not sure do we really need set the flag to INTERNAL_QUERY_FROM_EXECUTIL here. 

When catalog query is executed , it need set the flag to INTERAL_QUERY_FROM_EXEUTIL to bypasss checks on metadata tables. Here it looks we can skip set the flag.

Comment (arvind-narain) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: You are right - we are not directly accessing the metadata here - we can skip setting the parser flag.

**Comment by traflm** [comment]
Does review comments are fixed in the new commit?
If so, is this good to be merged?

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Blob.java (hunk 6 above)
Comment (DaveBirdsall) [inline core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/TrafT4Blob.java]: That's fine with me.

**Comment by traflm** [comment]
looks great to me, a big contribution. I will merge it since there are no more comments.

**Comment by arvind-narain** [comment]
Thanks once again for this great contribution.

**Pull request merged**